packages/edit-post/src/components/visual-editor/index.js

```diff
@@ -195,11 +206,25 @@ export default function VisualEditor( { styles } ) {
 		return { type: 'default' };
 	}, [ isTemplateMode, themeSupportsLayout, defaultLayout ] );
 
-	const blockListLayoutClass = classnames( {
-		'is-layout-constrained': themeSupportsLayout,
-		'is-layout-flow': ! themeSupportsLayout,
-		'has-global-padding': useRootPaddingAwareAlignments,
-	} );
+	const templateBlocks = parse( templateContent );
+	const postContentBlock = findPostContent( templateBlocks );
+	const postContentLayoutClasses = useLayoutClasses(
+		postContentBlock?.attributes?.layout,
+		defaultLayout?.definitions
+	);
+
+	const blockListLayoutClass = classnames(
+		{
+			'is-layout-flow': ! themeSupportsLayout,
+			'wp-container-visual-editor': themeSupportsLayout,
+		},
+		postContentLayoutClasses
+	);
+
+	const postContentLayoutStyles = useLayoutStyles(
+		postContentBlock,
+		'.wp-container-visual-editor'
+	);
```

Comment:
Will we ultimately want to wrap this in a `useMemo` so that we only run it when `templateContent` changes?

Reply:
Good question! I never know when it's best to `useMemo` or not 😅  Another option would be to return `null` from useLayoutStyles if there's no block.

Reply:
> Another option would be to return null from useLayoutStyles if there's no block.

Actually I don't think we can do that because it's using hooks underneath 😕

Reply:
> I never know when it's best to useMemo or not 😅 

Me either, it's often a bit tricky to work out when they're useful. In principle if we're doing potentially expensive things like block parsing on a component that might get rendered multiple times, I think it makes sense to add a `useMemo`. If I add a wrapper `findPostContentWrapper` function that calls `findPostContent` and log out the number of times it's called, in my local environment it looks like we call it 9 times (not sure how valid that is, since there's only 5 log lines there, but it at least suggests we might benefit from the useMemo caching 🤔):

<img width="224" alt="image" src="https://user-images.githubusercontent.com/14988353/191167470-d40e2f72-037f-4e90-9e6c-4b473df1b955.png">

From memory we can't call a hook from within another hook, so the optimisation might only be to wrap the block containing the two lines `parse` and `findPostContent` in a useMemo, so that we only parse / find blocks when the template content changes?

But this can totally be something to look at in the final stages of the PR.

Reply:
Ok I tried what you recommended and it seems to be working well! We can still see a jump in layout when the page first loads, I think because the initial data from `useSelect` takes a while to appear, but I don't think there's much to be done about that.

Reply:
> I think because the initial data from useSelect takes a while to appear, but I don't think there's much to be done about that.

Probably outside the scope of this PR, but I wonder if there's a way to preload the data for this selector? I see that the Navigation block has a preloading function [here](https://github.com/WordPress/gutenberg/blob/d5915916abc45e6682f4bdb70888aa41e98aa395/packages/edit-navigation/src/utils/index.js#L16) and sets apiFetch to use it [here](https://github.com/WordPress/gutenberg/blob/e102e3f853ea2fc15d976e5a633b1c0668beed5d/lib/experimental/navigation-page.php#L168). There's similar logic in gutenberg_initialise_editor [here](https://github.com/WordPress/gutenberg/blob/c1cf175c962ac2239020b80fb6a02341f7205f0b/lib/experimental/editor-settings.php#L47), so I wonder if there's a PHP filter of some kind to add in a desired API endpoint to pre-fetch or something? Might be a bit of a rabbithole 😅 

I don't at all understand the logic there, but just in case something along those lines sparks any ideas 🙂